// n8n-nodes-whatsable/dist/credentials/WhatsAbleApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhatsAbleApi = void 0;
const constants_1 = require("../shared/constants");
class WhatsAbleApi {
    constructor() {
        this.name = 'whatsAbleApi';
        this.displayName = 'WhatsAble API';
        this.documentationUrl = 'https://docs.whatsable.app/n8n-overview';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: `${constants_1.BASE_DOMAIN}/api:gncnl2D6`,
                url: '/check-api-key-across-projects',
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                },
            },
            rules: [
                {
                    type: 'responseSuccessBody',
                    properties: {
                        key: 'success',
                        value: false,
                        message: 'Invalid API key',
                    },
                },
            ],
        };
    }
}
exports.WhatsAbleApi = WhatsAbleApi;
