// n8n-nodes-whatsable/dist/credentials/WhatsAbleNotifyerApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhatsAbleNotifyerApi = void 0;
const constants_1 = require("../shared/constants");
class WhatsAbleNotifyerApi {
    constructor() {
        this.name = 'whatsAbleNotifyerApi';
        this.displayName = 'WhatsAble Notifyer System Trigger API';
        this.documentationUrl = 'https://docs.whatsable.app/n8n-overview';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
                required: true,
                description: 'API Key for WhatsAble',
            },
            {
                displayName: 'Webhook URL',
                name: 'productionWebhookUrl',
                type: 'string',
                default: '',
                required: true,
                description: 'Enter your webhook URL for testing. In production, this will be automatically determined.',
            }
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: `${constants_1.BASE_DOMAIN}/api:dBShrB6H`,
                url: '/n8n',
                method: 'POST',
                body: {
                    hookUrl: '={{$credentials.productionWebhookUrl}}'
                },
                headers: {
                    'Content-Type': 'application/json'
                }
            },
            rules: [
                {
                    type: 'responseSuccessBody',
                    properties: {
                        key: 'success',
                        value: false,
                        message: 'Invalid API key',
                    },
                },
            ],
        };
    }
}
exports.WhatsAbleNotifyerApi = WhatsAbleNotifyerApi;
