// n8n-nodes-aiscraper/dist/credentials/AIScraperApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AIScraperApi = void 0;
class AIScraperApi {
    constructor() {
        this.name = 'aiScraperApi';
        this.displayName = 'AI Scraper API';
        this.documentationUrl = 'https://docs.parsera.org/api/getting-started/';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
                required: true,
                description: 'Your Parsera API Key',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-API-KEY': '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://agents.parsera.org/v1',
                url: '/list',
            },
        };
    }
}
exports.AIScraperApi = AIScraperApi;
